### DISCUSSION TITLE
Stub Math Random

### DISCUSSION THREAD

**Discussion by Tsourdox** [cypress-io/cypress#29692] [Questions and Help]
How can I stub the math random function to predictably assert an outcome?

cy.stub(Math, "random").returns(0.8);

This doesn't seem to work.

**Comment by portyu9** [comment]
Hi @Tsourdox 

For Component Testing, make sure `Math.random` is stubbed **before the component executes the code that uses it**.

For example:

```js
cy.stub(window.Math, 'random').returns(0.8)

cy.mount(<MyComponent />)
```

Then application code such as:

```js
const value = Math.random()
```

will receive:

```text
0.8
```

The ordering matters. This will be too late if mounting immediately calls `Math.random()`:

```js
cy.mount(<MyComponent />)

cy.stub(window.Math, 'random').returns(0.8)
```

There is another case to check: if the random value is calculated at **module import time**, for example:

```js
const randomValue = Math.random()

export function getValue() {
  return randomValue
}
```

then even stubbing immediately before `cy.mount()` can be too late, because the imported module was already evaluated before the test started.

Prefer calculating it when the component/function runs:

```js
export function getValue() {
  return Math.random()
}
```

or, even better for deterministic code, inject the random-number function:

```js
export function getValue(random = Math.random) {
  return random()
}
```

Then the pure logic can be tested directly:

```js
expect(getValue(() => 0.8)).to.eq(/* expected result */)
```

For the Cypress component itself, though, this should be enough when `Math.random()` is called during/after mount:

```js
cy.stub(window.Math, 'random')
  .returns(0.8)
  .as('random')

cy.mount(<MyComponent />)

cy.get('@random').should('have.been.called')
```

Cypress automatically restores `cy.stub()` stubs between tests, so you do not need manual cleanup.

Please mark this answer as accepted if it helped, thank you.
